### ISSUE TITLE
Landfall decay issue in TC synthetic tracks

### ISSUE THREAD

**Issue by bguillod** [CLIMADA-project/climada_python#151]
I have found an issue with the decay of wind speed and pressure at landfall in the `climada.hazard.tc_track_synth` module. I here describe the issue as well as a part of the solution, with some open questions remaining. It consists of two interrelated problems and I take the time to go through in detail below, because I consider this a critical bug that requires a robust solution - thanks in advance for going through the full description!

I would be glad if **CLIMADA's Tropical Cyclone experts can have a look and give their opinion** if they have any.

I am preparing a pull request to solve these issues, but would be glad to get feedback **ASAP** so I can include that and make some critical choices.

### General description
Here is a reproducible example:
```
from climada.hazard.tc_tracks import TCTracks

sid = '1997018S11059'
tracks = TCTracks()
tracks.read_ibtracs_netcdf(provider='usa', storm_id=sid, estimate_missing=True)

tracks.equal_timestep(3)
tracks.calc_perturbed_trajectories(nb_synth_tracks=10)
tracks.plot()
```
![image](https://user-images.githubusercontent.com/13904718/108193165-740a0e00-7115-11eb-9635-e7ab95ee7fa9.png)

Let me take the 3rd synthetic as it illustrates both issues: It makes a first landfall over Madagascar, then a 2nd landfall over Mozambique, and then goes back over the ocean, where it ends:
```
track_1 = TCTracks()
track_1.data = [tracks.get_track('1997018S11059_gen3')]
track_1.plot()
```
![image](https://user-images.githubusercontent.com/13904718/108193283-91d77300-7115-11eb-8c47-db43daef4352.png)

Let me first display a time series of the track, using the following functions:
```
import numpy as np
import pandas as pd
from matplotlib.dates import date2num
from climada.hazard.tc_tracks import TCTracks
import matplotlib.pyplot as plt

def get_lf_df(tr_df):
    lf_start = tr_df.iloc[np.where(np.diff(tr_df['on_land'].astype(float)) == 1)[0],:]['time']
    lf_end = tr_df.iloc[np.where(np.diff(tr_df['on_land'].astype(float)) == -1)[0],:]['time']
    if len(lf_end) < len(lf_start):
        lf_end = lf_end.append(max(tr_df['time']))
    lf_dat = pd.DataFrame({'start_lf':lf_start, 'end_lf': lf_end}).reset_index(drop=True)
    return lf_dat


def get_tracks_df(tracks, sid_hist, ens_nb):
    # get historical track df
    tr_hist = TCTracks()
    tr_hist.data = [tracks.get_track(sid_hist)]
    tr_hist_df = tr_hist.to_geodataframe(as_points=True)
    # get synthetic track df
    tr_synth = TCTracks()
    tr_synth.data = [tracks.get_track(sid_hist+'_gen'+str(ens_nb))]
    tr_synth_df = tr_synth.to_geodataframe(as_points=True)
    return tr_synth_df, tr_hist_df

def plot_landfall_data(tracks, sid_hist, ens_nb):
    tr_df, tr_hist_df = get_tracks_df(tracks, sid_hist, ens_nb)
    fig, ax = plt.subplots(figsize=(10,6))
    lf_dat = get_lf_df(tr_df)
    lf_hist = get_lf_df(tr_hist_df)
    #     ws = ax.scatter(tr_df['time'],tr_df['max_sustained_wind'])
    ws = ax.plot(tr_df['time'],tr_df['max_sustained_wind'], marker='x')
    wo = ax.scatter(tr_hist_df['time'], tr_hist_df['max_sustained_wind'], marker='+', color='red')
    # ax = pd.DataFrame(t1_syn_df).plot(x='time',y='max_sustained_wind')
    for i in range(lf_hist.shape[0]):
        lfo = ax.axvspan(date2num(lf_hist['start_lf'][i]), date2num(lf_hist['end_lf'][i]), 
                   label="Landfalls",color="grey", alpha=0.3)
    for i in range(lf_dat.shape[0]):
        lfs = ax.axvspan(date2num(lf_dat['start_lf'][i]), date2num(lf_dat['end_lf'][i]), 
                   label="Landfalls",color="green", alpha=0.3)
    ax.set_title('Maximum sustained wind for cyc_id %s' % sid_hist+'_gen'+str(ens_nb))
    ax.set_ylabel('kts')
    legend1 = plt.legend((ws[0], wo), ('Actual', 'Historical track'), loc=2)
    ax.legend((lfs, lfo), ('Landfall', 'Landfall (historical)'), loc=3)
    plt.gca().add_artist(legend1)

    fig, ax = plt.subplots(figsize=(10,6))
    ws = ax.plot(tr_df['time'],tr_df['central_pressure'], marker='x')
    wo = ax.scatter(tr_hist_df['time'], tr_hist_df['central_pressure'], marker='+', color='red')
    we = ax.scatter(tr_hist_df['time'], tr_hist_df['environmental_pressure'], marker='.', color='orange')
    #     wes = ax.scatter(tr_df['time'], tr_df['environmental_pressure'], marker='.', color='green')
    # ax = pd.DataFrame(t1_syn_df).plot(x='time',y='max_sustained_wind')
    for i in range(lf_hist.shape[0]):
        ax.axvspan(date2num(lf_hist['start_lf'][i]), date2num(lf_hist['end_lf'][i]), 
                   label="Landfalls",color="grey", alpha=0.3)
    for i in range(lf_dat.shape[0]):
        ax.axvspan(date2num(lf_dat['start_lf'][i]), date2num(lf_dat['end_lf'][i]), 
                   label="Landfalls",color="green", alpha=0.3)
    ax.set_title('Central pressure for cyc_id %s' % sid_hist+'_gen'+str(ens_nb))
    ax.set_ylabel('hPa')
    ax.legend((ws[0], wo, we), ('Actual', 'Historical track', 'Environmental pressure'))
```
And then plot the track time series of wind speed and central & environmental pressure with:
```
plot_landfall_data(tracks, sid, 3)
```
![image](https://user-images.githubusercontent.com/13904718/108204413-10d3a800-7124-11eb-9e1b-2f52d42cc2bc.png)
![image](https://user-images.githubusercontent.com/13904718/108204424-13ce9880-7124-11eb-8be4-191d0871c377.png)
These plots show the synthetic track values (blue) as well as the original historical track (red). Landfalls in the synthetic track are green background, while the (single) landfall of the historical track is in grey. The central pressure plot also shows environmental pressure (yellow), which is the same in both historical and synthetic tracks.

The issues are the following:
1. **The synthetic tracks values after the last landfall are as they were in the historical track** (see especially for wind speed, which creates a substantial jump for the last few values).
2. Central pressure for during the 2nd landfall is as in the historical track, giving the impression that **the landfall of the synthetic track generates a strong increase in pressure drop**. This clearly is also unrealistic.

### ISSUE 1: NO CORRECTION AFTER LANDFALL IF THE TRACK ENDS ON THE OCEAN
The reason for issue 1 is the following line
https://github.com/CLIMADA-project/climada_python/blob/038cfc40ebb83c6f114557ed7576bbba022e9860/climada/hazard/tc_tracks_synth.py#L744
which states that values after landfall are only corrected if there is a subsequent landfall.

I propose to fix this by replacing these lines
https://github.com/CLIMADA-project/climada_python/blob/038cfc40ebb83c6f114557ed7576bbba022e9860/climada/hazard/tc_tracks_synth.py#L744-L753
with this
```
        # correct values of sea after a landfall (until next landfall, if any)
        if land_sea < track.time.size:
            if idx + 1 < sea_land_idx.size:
                # if there is a next landfall, correct until last point before
                # reaching land again
                end_cor = sea_land_idx[idx + 1]
            else:
                # if there is no further landfall, correct until the end of
                # the track
                end_cor = track.time.size
            rndn = 0.1 * float(np.abs(np.random.normal(size=1) * 5) + 6)
            r_diff = track.central_pressure[land_sea].values - \
                     track.central_pressure[land_sea - 1].values + rndn
            track.central_pressure[land_sea:end_cor] += - r_diff

            rndn = rndn * 10  # mean value 10
            r_diff = track.max_sustained_wind[land_sea].values - \
                     track.max_sustained_wind[land_sea - 1].values - rndn
            track.max_sustained_wind[land_sea:end_cor] += - r_diff

```
This change leads to the following time series for wind speed:
![image](https://user-images.githubusercontent.com/13904718/108194857-65245b00-7117-11eb-9a35-865986a43fe9.png)
which to me seems to make much more sense.

The only thing I don't like is that just after the first landfall, the historical track was itself making landfall, leaving to decreasing wind speed (and increasing pressure), which is then also applied to the synthetic track. But I do not see how to correct for this without explicitly modeling the potential increase in TC intensity once it is back to the ocean.

Also, note that I did not touch the formulation of this correction, just when it should be applied. **The formulation itself could probably be improved but is out of the scope of this issue/PR**.

### ISSUE 2: CENTRAL PRESSURE NOT CORRECTED DURING LANDFALL IN SOME CASES
The reason for the second issue is a combination of how S is calculated
https://github.com/CLIMADA-project/climada_python/blob/038cfc40ebb83c6f114557ed7576bbba022e9860/climada/hazard/tc_tracks_synth.py#L729
and that decay is only applied if it increases central pressure (which in principle makes sense):
https://github.com/CLIMADA-project/climada_python/blob/038cfc40ebb83c6f114557ed7576bbba022e9860/climada/hazard/tc_tracks_synth.py#L732-L733

S (output of `_calc_decay_ps_value`) is the ratio `between the environmental pressure at the last point still on land of this landfall` (or last point of track if track doesn't go back to ocean) and `central pressure at last point before landfall`. In this example, at the time this line is run (for the 2nd landfall) central pressure is larger than environmental pressure. We don't see this in the plot because later on it is set to environmental pressure if larger than it:
https://github.com/CLIMADA-project/climada_python/blob/038cfc40ebb83c6f114557ed7576bbba022e9860/climada/hazard/tc_tracks_synth.py#L757-L758
Hence, `S < 1`, meaning that the decay would lead to a decrease in pressure (instead of an increase. Therefore, it is not applied, leaving central pressure as it was in the historical track.
The undesirable result, however, is that central pressure strongly decreases at this 2nd landfall.

A first step to fix this is to set `central_pressure` to `environmental_pressure` within the loop of landfalls, i.e. add the following after the `# correct values of sea after a landfall (until next landfall, if any)` code block:
```
            np.warnings.filterwarnings('ignore')
            cor_p = track.central_pressure.values > track.environmental_pressure.values
            track.central_pressure[cor_p] = track.environmental_pressure[cor_p]
            track.max_sustained_wind[track.max_sustained_wind < 0] = 0
```

Doing this leads to the following
![image](https://user-images.githubusercontent.com/13904718/108205273-1f6e8f00-7125-11eb-80a7-de6175448f25.png)
which is already much better. Still, there is a small drop in `central_pressure` just after the landfall (and perhaps in some other cases that drop could be larger, notably if the historical `central_pressure` value just after the landfall is much lower than at the end of the landfall in the synthetic track. Let's examine again how `central_pressure` is corrected after landfall:
https://github.com/CLIMADA-project/climada_python/blob/038cfc40ebb83c6f114557ed7576bbba022e9860/climada/hazard/tc_tracks_synth.py#L745-L748
If we ignore `r_diff` (generally small), the difference between the first point after landfall (historical track value) and the last landfall point (after applying landfall decay) is applied, which in some cases might create a jump.
For instance, consider the following track (2nd ensemble member of another historical track):
```
sid = '1988021S12080'
tracks_2 = TCTracks()
tracks_2.read_ibtracs_netcdf(provider='usa', storm_id=sid, estimate_missing=True)
tracks_2.calc_perturbed_trajectories(nb_synth_tracks=10)
plot_landfall_data(tracks_2, sid, 1)
```
Here are the plot (including changes mentioned above):
![image](https://user-images.githubusercontent.com/13904718/108207523-28ad2b00-7128-11eb-9580-fb1ab3650659.png)
![image](https://user-images.githubusercontent.com/13904718/108207528-2c40b200-7128-11eb-95f6-d4aa9cb21084.png)

As you can see, wind speed after the (only) landfall keeps is corrected, but central_pressure is not because environmental pressure (and hence also the capped central pressure) is higher at the start of the landfall than it is at the end. To fix this, I suggest that if S<=1, central_pressure is set to environmental pressure during the whole landfall by adding the following just after `p_decay = _calc_decay_ps_value(track, p_landfall, land_sea - 1, s_rel)`
```
        if p_decay <= 1:
            track.central_pressure[sea_land:land_sea] = track.environmental_pressure[sea_land:land_sea]
        else:
            p_decay = _decay_p_function(p_decay, p_rel[ss_scale_idx][1],
                                        track.dist_since_lf[sea_land:land_sea].values)
            # dont apply decay if it would decrease central pressure
            p_decay[p_decay < 1] = track.central_pressure[sea_land:land_sea][p_decay < 1] / p_landfall
            track.central_pressure[sea_land:land_sea] = p_landfall * p_decay
```
which leads to the same wind but the following central pressure evolution:
![image](https://user-images.githubusercontent.com/13904718/108208492-665e8380-7129-11eb-8d83-5a978ee2696e.png)

This seems to fix the issue. Does it also make sense physically?

### CONCLUSION

I propose some fixes but am open to other suggestions. Also, I haven't tested these fixes for many cases, so if someone would like to (or has a similar case they'd like me to try out), please do.

**Comment by bguillod** [comment]
@chahank @tovogt I assigned both of you to this issue, but perhaps there are other people working on TCs that might have some input to this? If so, please assign them :)

**Comment by tovogt** [comment]
Thanks for taking the time to elaborate on this question! Your description is very clear and I appreciate the style a lot.

Still, I won't go into this because I think that this is actually a research question and not a question of implementation or algorithmic design. More precisely, there is a fundamental flaw in the `calc_perturbed_trajectories` functionality of CLIMADA: it's unpublished and has never been validated, as far as I know. There's not even anyone working on a validation. The design of this functionality is completely "ad hoc", without even any climatological or meteorological justification. That's why I'm not at all surprised that you find fundamental flaws in the output.

Since the whole functionality is unpublished and lacks justification as well as validation, it remains completely unclear what this functionality claims to achieve and which purpose it serves. There's no clear use-case given anywhere in the documentation, as far as I know. Nothing is said about the limitations etc. And that's why I can't really tell whether the behavior you describe in this issue qualifies as a flaw in the first place. The function doesn't claim anywhere not to show the kind of behavior you describe. It's unlikely, but maybe it's intended?

**Comment by bguillod** [comment]
@tovogt I understand your point and I am fine if you don't want to go into this, but I only partially agree on the reasons. Until someone works on the research question part of it (which I am not aware is planned at the moment), I think it is useful to fix the flaws I describe. This functionality is used in many cases, including in several companies, so I think it is clearly worth fixing - but if CLIMADA's "core" team doesn't want these fixes in the "official" CLIMADA and prefer to develop a more robust random walk methodology I am fine with it, in which case we don't necessarily need to merge them and can keep on a separate branch.

**Comment by tovogt** [comment]
Yes, this is my perspective which is the perspective of someone who is neither part of the CLIMADA core team nor a user of this function.

**Comment by chahank** [comment]
Thank you both for your input. Thank you in particular @bguillod  for making such a detailed explanation. I must say I tend to agree with @tovogt on the fact that the synthetic track method in CLIMADA is not on a scientifically satisfying level, and I guess this implies it is also questionable that it is used for commercial purposes. This function was not intended to be a real alternative to tropical cyclone track modeling. Hence, I think we should take up the fundamental question of what the synthetic track method is meant to achieve. 

For the moment, I do not see any reason for @bguillod not to improve the current version, so thanks for the pull request! However, a long-term reflection is probably needed. Bear in mind though that we are not tropical cyclone modelers.

**Comment by bguillod** [comment]
Thanks for both of you, too. @chahank I have replied to your comments, let me know if anything is unclear.
I think we need to be clear that the synthetic tracks approach is a simple statistical model, not an explicitly tropical cyclone model, so in that sense I of course agree with both of you. But whether this approach is sufficient depends on the context and what one is trying to achieve. I certainly would welcome a more robust approach in CLIMADA if that comes at some point.

**Comment by tovogt** [comment]
Since we had a discussion here about the method being published or not, here is a reference that actually describes the method as it was back then:

Gettelman, A., Bresch, D.N., Chen, C.C., Truesdale, J.E., Bacmeister, J.T. (2018): Projections of future tropical cyclone damage with a high-resolution global climate model. Climatic Change 146(3–4): 575–585. (http://link.springer.com/10.1007/s10584-017-1902-7)

However, the method is not evaluated or validated in the paper. There is a short description of the method in the supplementary material of the paper:

> This is a method for interpolation from a few tracks to a regular track density field. The base input cyclone tracks can come from observations or GCM simulations. Starting from a slightly perturbed location compared to the original track at t=0, each probabilistic track deviates from its original by following a directed random walk, ‘guided’ by the original track, and with a perturbation to the wind speed randomly selected from a normal distribution with mean of the actual speed and one standard deviation from all the cyclone initial speeds. The area affected away from the center of the track is determined by scaling relationships for the size of storms (Bresch, 2014). This way, a probabilistic set of artificial tracks is created which preserves, as
long as the directed random walk does not veer too far away from the original, the general characteristics of the original set, yet provides a much more robust basis for landfall statistics (by giving a more robust estimate of the probability distribution of storms that would result from a given set of conditions) and hence damage calculations (Bresch, 2014).

By (Bresch, 2014), the author is referring to the (unpublished) documentation of the old MATLAB CLIMADA software package.

@bguillod Can you maybe summarize in how far this description still applies to the current implementation, or what might have changed?

**Comment by bguillod** [comment]
@tovogt I do not see the `perturbation to the wind speed randomly selected from a normal distribution with mean of the actual speed and one standard deviation from all the cyclone initial speeds` in the code, except perhaps the recovery after landfall. Besides that wind speed (and pressure) are not altered.
I think the description you wrote is not very accurate.

**Comment by bguillod** [comment]
Issue solved in PR #152
